### PR TITLE
ft-user-signin:user should be able to sign in

### PR DESCRIPTION
**Pull request by athanase-matabaro**
# PR Details
user should be able to sign in 

## Description
- upgrade dependecies
- create validator 
- new endpooint to allow the user sign in
- new services at the data base layer 
- controller logic for signing up the user 
- user model updated 
- user migration updated 
- json web token helper modified
- routing system updated 

## Motivation and Context
this changes have been required because the intialisation of the second endpoint result into many error when running the app so for the express server to run correctly the above changes was crutial.

## How Has This Been Tested
using postman to intiate the request and laragon for the database environment 

## Types of changes

- [x ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x ] All new and existing tests passed.

### CODE REVIEW

File: src/app/helpers/jwt.js [hunk 1]
```diff
@@ -9,9 +10,8 @@ class Token {
 
   static createToken(
     data = {},
-    expireIn = { expireIn: '86400' },
   ) {
-    const token = jwt.sign({ data }, process.env.SECRET_KEY, expireIn);
+    const token = jwt.sign({ data }, process.env.SECRET_KEY);
```
Comment (chrismatgit) [inline src/app/helpers/jwt.js:14]: Why did you remove the expiring time?

File: src/app/modules/authController.js [hunk 2]
```diff
@@ -6,6 +6,22 @@ import jwt from '../helpers/jwt';
 
 class UserController {
   static async signup(req, res) {
+    const checkEmail = req.body.email;
+    const checkUser = req.body.username;
```
Comment (chrismatgit) [inline src/app/modules/authController.js]: I can suggest cleaning your code by destructuring from here.
```
const { username, email, password } = req.body
```
Then you can pass them into emailUsed and the other one instead of creating checkEmail & checkUser variables.

File: src/app/modules/authController.js [hunk 3]
```diff
@@ -6,6 +6,22 @@ import jwt from '../helpers/jwt';
...
+      console.log(resp);
+      return resp;
+    }
+    if (usernUsed) {
+      const resp = GenericHandler.error(res, 409, 'username exist already');
+      console.log(resp);
+      return resp;
+    }
+
     const { email, username, password } = req.body;
```
Comment (chrismatgit) [inline src/app/modules/authController.js:9]: Destructure from the beginning of the method.

File: src/app/modules/authController.js [hunk 4]
```diff
@@ -14,16 +30,42 @@ class UserController {
...
     const result = await UserServices.create(authData);
-    const { userId } = result.dataValues;
-    const token = jwt.createToken(userId);
 
+    const uID = req.body.username;
+    const getId = await UserServices.findUserid(uID);
+
+    const { id } = getId;
+    const token = jwt.createToken(id);
     if (!result) {
```
Comment (chrismatgit) [inline src/app/modules/authController.js]: In your code you are just passing an object containing the password. You should consider storing all the information into the database not only the password. So after creating a user, the result is stored into the the variable **result**. Then you can use that object to generate the token because it should contain all those information. Then you can destructure those information to get the **id, username & email**. Clean your code...

File: src/app/modules/authController.js [hunk 5]
```diff
@@ -14,16 +30,42 @@ class UserController {
       password: hashPassword,
     };
     const result = await UserServices.create(authData);
-    const { userId } = result.dataValues;
-    const token = jwt.createToken(userId);
 
+    const uID = req.body.username;
+    const getId = await UserServices.findUserid(uID);
```
Comment (chrismatgit) [inline src/app/modules/authController.js]: Don’t need to find the **userId** by making the request. Destruct the **result**

File: src/app/modules/authController.js [hunk 6]
```diff
@@ -14,16 +30,42 @@ class UserController {
...
+      return resp;
+    }
+
+    const realPass = Password.passCompare(
+      password,
+      getUser.password,
+    );
+
+    if (!getUser || !realPass) {
+      GenericHandler.error(res, 404, 'incorrect credentials');
```
Comment (eliezerbasubi) [inline src/app/modules/authController.js]: This looks good to me. Just one thing to improve.
- Create a file that holds all the response statuses and response messages.

File: src/app/modules/authController.js [hunk 7]
```diff
@@ -14,16 +30,42 @@ class UserController {
...
   }
+
+  static async signin(req, res) {
+    const { username, password } = req.body;
+    const checkUser = req.body.username;
+    const getUser = await UserServices.signinUser(checkUser);
+
+    if (!getUser) {
+      const resp = GenericHandler.error(res, 404, 'user does not exist');
+      console.log(resp);
```
Comment (chrismatgit) [inline src/app/modules/authController.js]: Remove all the log. They are used in the debugging but not on the production

File: src/app/modules/authController.js (hunk 6 above)
Comment (athanase-matabaro) [reply src/app/modules/authController.js]: okay thanks 
but can i do after or before merging ?

Comment (eliezerbasubi) [reply src/app/modules/authController.js]: First, make the changes then merge.

File: src/app/routes/index.js [hunk 8]
```diff
@@ -3,17 +3,20 @@ import GenericHandler from '../helpers/responses';
 import authRoutes from './auth/userRouters';
 
 const apiVersion = '/api/v1';
```
Comment (chrismatgit) [inline src/app/routes/index.js:5]: Put this information into the **.env** file

**Review by chrismatgit — APPROVED** [review]
Looks good. Just work on the feedback before merging.

File: src/app/modules/authController.js (hunk 6 above)
Comment (athanase-matabaro) [reply src/app/modules/authController.js]: okay got it

File: src/app/helpers/jwt.js (hunk 1 above)
Comment (athanase-matabaro) [reply src/app/helpers/jwt.js:14]: it was sending an error like "exprireIn" is not allowed in option

Comment (chrismatgit) [reply src/app/helpers/jwt.js:14]: Because you did not import the env in that file. Try and then it will work.

File: src/app/modules/authController.js (hunk 2 above)
Comment (athanase-matabaro) [reply src/app/modules/authController.js]: i've been destructuring separetly because i needed a specifiic element of the request body that i could used in function defined under services and the variables was to avoid duplication but okay

File: src/app/modules/authController.js (hunk 3 above)
Comment (athanase-matabaro) [reply src/app/modules/authController.js:9]: okay thanks

File: src/app/modules/authController.js (hunk 4 above)
Comment (athanase-matabaro) [reply src/app/modules/authController.js]: everything is being stored in the database , maybe just that the token should be generated based on whole result

File: src/app/modules/authController.js (hunk 5 above)
Comment (athanase-matabaro) [reply src/app/modules/authController.js]: okay thanks

File: src/app/routes/index.js (hunk 8 above)
Comment (athanase-matabaro) [reply src/app/routes/index.js:5]: okay got it

File: src/app/modules/authController.js (hunk 7 above)
Comment (athanase-matabaro) [reply src/app/modules/authController.js]: actually i was using the next(); function, but it was returning it as undefined
any suggestion will be welcomed

File: src/app/helpers/jwt.js (hunk 1 above)
Comment (athanase-matabaro) [reply src/app/helpers/jwt.js:14]: okay thanks
